text/0032-dust-protection/0032-dust-protection.md

```diff
@@ -0,0 +1,179 @@
+  </tr>
+  <tr>
+    <td>Amount</td>
+    <td>uint64</td>
+    <td>The amount of tokens to deposit with this <i>SigLockedDustAllowanceOutput</i> output.</td>
+  </tr>
+</table>
+
+
+### Validation
+
+Let A be the address that should hold the dust outputs' balances. Let S be the sum of all the amounts of all unspent `SigLockedDustAllowanceOutputs` on A. Then, the maximum number of allowed dust outputs on A is S divided by 10,000 and rounded down, i.e. 100 outputs for each 1 Mi deposited.
+
+The amount of a `SigLockedDustAllowanceOutput` must be at least 1 Mi. Apart from this, `SigLockedDustAllowanceOutputs` are processed identical to `SigLockedSingleOutput`. The transaction validation as defined in [Draft RFC-18](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md), however, needs to be adapted.
+
+_Syntactical validation_ for `SigLockedDustAllowanceOutput`:
+- The `Address` must be unique in the set of `SigLockedDustAllowanceOutputs` in one transaction T. However, there can be one `SigLockedSingleOutput` and one `SigLockedDustAllowanceOutputs` T.
+- The `Amount` must be ≥ 1,000,000.
+
+The _semantic validation_ remains unchanged and are checked for both `SigLockedSingleOutputs` and `SigLockedDustAllowanceOutput`, but this RFC introduces one additional criterion:
+
+A transaction T
+  - consuming a `SigLockedDustAllowanceOutput` on address A **or**
+  - creating a dust output with address A,
+
+is only semantically valid, if, after T is booked, the number of unspent dust outputs on A does not exceed the allowed threshold of S / 10,000.
+
+The following pseudocode demonstrates the new rules:
+
+```python
+boolean DustValidation(Transaction tx):
+    # maps return 0 for non-existent keys
+    dustAllowanceAddressesToBalances = new map()
+    dustAddressesToNumOfDustOutputs = new map()
+    
+    if tx not null:
+        for output in tx.outputs:
+            if output.type == SigLockedDustAllowanceOutput:
+                # must be at least 1 Mi
+                if output.balance < 1_000_000:
+                    return false
+                # only one allowance deposit per address per transaction
+                if (dustAllowanceAddressesToBalances[output.address] == 0):
+                    dustAllowanceAddressesToBalances[output.address] = output.amount
+                else
+                    return false
+            # count new dust outputs
+            if output.type == SigLockedSingleOutput & output.balance < 1_000_000:
+                dustAddressesToNumOfDustOutputs[output.address] += 1
+        for input in tx.inputs:
+            # load from the UTXO set
+            output = UTXO.loadUtxoForInput(input)
+            if output.amount < 1_000_000:
+               dustAddressesToNumOfDustOutputs[output.address] -= 1
+            if output.type == SigLockedDustAllowanceOutput:
+                dustAllowanceAddressesToBalances[output.address] =- output.amount
```

Comment:
...but here you assume that some initial deposit could be there, consumed in the input. You may therefore ending up with a value mapped for that deposit < 0. Take the case that you have a Tx that has an input SigLockedDustAllowanceOutput UTXO of 1.1Mi and a deposit output UTXO of 1.0Mi. The value in this map for this `output.address` key will be -0.1Mi.
You probably took this into account, but I find it very confusing and error-prone, especially because on line 139 and 141 you could end up with a negative number!